src/Dapr.Client/InvocationHandler.cs

```diff
@@ -0,0 +1,137 @@
+        public InvocationHandler()
+        {
+            this.parsedEndpoint = new Uri(GetDefaultDaprEndpoint(), UriKind.Absolute);
+            this.apiToken = Environment.GetEnvironmentVariable("DAPR_API_TOKEN");
+        }
+
+        /// <summary>
+        /// Gets or the sets the URI of the Dapr HTTP endpoint used for service invocation.
+        /// </summary>
+        /// <returns>The URI of the Dapr HTTP endpoint used for service invocation.</returns>
+        public string DaprEndpoint
+        {
+            get
+            {
+                return this.parsedEndpoint.OriginalString;
+            }
+            set
+            {
+                if (value == null)
+                {
+                    throw new ArgumentNullException(nameof(value));
+                }
+
+                // This will throw a reasonable exception if the URI is invalid.
+                this.parsedEndpoint = new Uri(value, UriKind.Absolute);
+            }
+        }
+
+        // Internal for testing
+        internal string? DaprApiToken
+        {
+            get => this.apiToken;
+            set => this.apiToken = value;
+        }
+
+        /// <inheritdoc />
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            var original = request.RequestUri;
+            if (!this.TryRewriteUri(request.RequestUri, out var rewritten))
+            {
+                throw new ArgumentException($"The request URI '{original}' is not a valid Dapr service invocation destination.", nameof(request));
+            }
+
+            try
+            {
+                if (this.apiToken is string)
+                {
+                    request.Headers.Add("dapr-api-token", this.apiToken);
+                }
+                request.RequestUri = rewritten;
+
+                return await base.SendAsync(request, cancellationToken);
+            }
+            finally
+            {
+                request.RequestUri = original;
+                request.Headers.Remove("dapr-api-token");
+            }
+        }
+
+        // Internal for testing
+        internal bool TryRewriteUri(Uri uri, [NotNullWhen(true)] out Uri? rewritten)
+        {
+            // For now the only invalid cases are when the request URI is missing or just silly.
+            // We may support additional cases for validation in the future (like an allow-list of App-Ids).
+            if (uri is null || !uri.IsAbsoluteUri || (uri.Scheme != "http" && uri.Scheme != "https"))
+            {
+                // do nothing
+                rewritten = null;
+                return false;
+            }
+
+
+            var builder = new UriBuilder(uri)
+            {
+                Scheme = this.parsedEndpoint.Scheme,
+                Host = this.parsedEndpoint.Host,
+                Port = this.parsedEndpoint.Port,
+                Path = $"/v1.0/invoke/{uri.Host}/method" + uri.AbsolutePath,
```

Comment:
Should "v1.0" be a const now that we are approaching v1?

Reply:
I don't think there's any value in trying to optimize for that. I'm **certain** that whatever we do for versioning post-1.0 will be more complex than just updating a constant.